akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala

```diff
@@ -71,64 +67,34 @@ class Worker private (
       workId: String): PartialFunction[(scaladsl.ActorContext[Worker.Message], Signal), Behavior[Worker.Message]] = {
     case (_, Terminated(_)) =>
       ctx.log.info("Work executor terminated. Reporting failure")
-      workManagerProxy ! WorkManager.WorkFailed(adapter, workId)
+      // FIXME no way to report yet, maybe timeout is okay?
       // need to re-create the work executor
       idle(createWorkExecutor())
   }
 
   def idle(workExecutor: ActorRef[ExecuteWork] = createWorkExecutor()): Behavior[Worker.Message] =
-    Behaviors.setup[Worker.Message] { ctx =>
       Behaviors.receiveMessagePartial[Worker.Message] {
-        case WorkIsReady =>
-          // this is the only state where we reply to WorkIsReady
-          workManagerProxy ! WorkerRequestsWork(workerId, adapter)
-          Behaviors.same
-
-        case SubmitWork(Work(workId, job: Int)) =>
-          ctx.log.info("Got work: {}", job)
-          workExecutor ! WorkExecutor.ExecuteWork(job, ctx.self)
-          working(workId, workExecutor)
-
-        case Ack(_) =>
-          Behaviors.same
-
+        case DeliveredMessage(confirmTo, message, seqNr) =>
+          message match {
+            case WorkManager.DoWork(w@Work(workId, job: Int)) =>
+              ctx.log.info("Got work: {}", w)
+              workExecutor ! WorkExecutor.ExecuteWork(job, ctx.self.narrow[Worker.WorkComplete])
+              working(workId, workExecutor, confirmTo, seqNr)
+          }
       }
-    }
 
-  def working(workId: String, workExecutor: ActorRef[ExecuteWork]): Behavior[Worker.Message] =
-    Behaviors.setup { ctx =>
+  def working(workId: String, workExecutor: ActorRef[ExecuteWork], confirmTo: ActorRef[ConsumerController.Confirmed], seqNr: Long): Behavior[Worker.Message] =
       Behaviors
         .receiveMessagePartial[Worker.Message] {
           case Worker.WorkComplete(result) =>
-            ctx.log.info("Work is complete. Result {}.", result)
-            workManagerProxy ! WorkManager.WorkIsDone(workerId, workId, result, adapter)
-            ctx.setReceiveTimeout(workAckTimeout, WorkTimeout)
-            waitForWorkIsDoneAck(result, workId, workExecutor)
-
-          case _: SubmitWork =>
-            ctx.log.warn("Yikes. WorkManager told me to do work, while I'm already working.")
+            ctx.log.info("Work is complete. Result {}. Sequence nr {}", result, seqNr)
+            confirmTo ! ConsumerController.Confirmed(seqNr)
+            idle(workExecutor)
+          case _: DeliveredMessage =>
+            ctx.log.warn("Yikes. Reliable delivery told me to do work, while I'm already working.")
             Behaviors.unhandled
 
         }
         .receiveSignal(reportWorkFailedOnRestart(workId))
```

Comment:
missing confirmation to the ConsumerController if WorkExecutor terminated, goes back to idle here